core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -609,9 +605,17 @@ public final boolean refreshNeeded() {
     }
 
     /**
+     * Register a listener that is called whenever the searcher is refreshed. See {@link SearcherManager#addListener(RefreshListener)}.
+     */
+    public final void registerSearchRefreshListener(RefreshListener listener) {
```

Comment:
I think we should only have one RefreshListener and it should be passed via `EngineConfig`